Vite experimental flag and changeset (#1189)
seek-oss/sku: .changeset/flat-papayas-suffer.md

---
'sku': minor
---

`vite` experimental support.

- Add experimental support for `vite` as a build tool.

## Usage

To enable `vite` as the build tool you will have to follow the steps below.

### Update sku.config.ts

First switch your project to `vite` by adding the following line to your `sku.config.ts`

```typescript
// sku.config.ts
import type { SkuConfig } from 'sku';

export default {
  bundler: 'vite', // sku will use vite as the bundler.
  ...
} satisfies SkuConfig;
```

### Migrating your client entrypoint

With `vite` the `index.html` will be generated with a different id for the root element.

> Make sure that your client entrypoint exports a function that hydrates the root element as the default export.

Reviewer: I assume this won't be necessary once `renderDocument` is supported in vite. Maybe worth noting that this is a temporary change.

```diff
import { hydrateRoot } from 'react-dom/client';
import { BrowserRouter } from 'react-router';

import { App } from './App';

export default ({ site }: { site: string }) => {
  hydrateRoot(
+    document.getElementById('root')!,
-    document.getElementById('app')!,
    <BrowserRouter>
      <App site={site} />
    </BrowserRouter>,
  );
};
```

### Migrating your render entrypoint.

The interface for the render entrypoint has seen some more significant changes.

- The `renderToString` function has been replaced with `renderToPipeableStream`.
- The `renderApp` function has been renamed to `render`.

> - The `SkuProvider` component has been removed from the `renderApp` function. Instead, you have to use the `LoadableProvider` from the `sku/vite/loadable` export.

Reviewer: Is `LoadableProvider` always necessary, or is it only useful if you're using `loadable`?

Author: Only needed if you use `loadable` which is what the SkuProvider did I believe? I'll mention it more explicitly.

- The `preloadAll` function has been added. Awaiting this function ensures that all `loadable` imported files will be preloaded before sku renders the page.

> - The `renderDocument` function has been removed. It's functionality will be added in a future update.

Reviewer: ```suggestion
- The `renderDocument` function has been removed. Its functionality will be added in a future update.
```

-        <StaticRouter location={route} context={{}}>
-          <App site={site} />
-        </StaticRouter>
-      </SkuProvider>,
-    );
-  },
+  render: async ({ options, renderContext, site, url }) => {
+     const { loadableCollector } = renderContext;
+
+     await preloadAll();
+
+     return renderToPipeableStream(
+       <LoadableProvider value={loadableCollector}>
+         <StaticRouter location={url} context={{}}>
+           <App site={site} />
+         </StaticRouter>
+       </LoadableProvider>,
+       options,
+     );
+  },

  provideClientContext: ({ site }) => ({
    site,
  }),

-  renderDocument: ({ app, bodyTags, headTags }) => {
-    return html/* html */ `
-      <!DOCTYPE html>
-      <html>
-        <head>
-          <meta charset="UTF-8" />
-          <title>hello-world</title>
-          <meta name="viewport" content="width=device-width, initial-scale=1" />
-          ${headTags}
-        </head>
-        <body>
-          <div id="app">${app}</div>
-          ${bodyTags}
-        </body>
-      </html>
-    `;
-  },
- } satisfies Render;
+ } satisfies ViteRender;
```

### Migrating from `sku/@loadable/components` to `sku/vite/loadable`

> The `sku/@loadable/components` package has been replaced with `sku/vite/loadable`.

Reviewer: ```suggestion
The `sku/@loadable/components` entrypoint has been replaced with `sku/vite/loadable`.
```

> The new `sku/vite/loadable` package relies on the `react <Suspense>` component for the loading fallback.

Reviewer: ```suggestion
The new `sku/vite/loadable` entrypoint relies on the React's [`<Suspense>`](https://react.dev/reference/react/Suspense) component for the loading fallback.
```

```diff
- import loadable from 'sku/@loadable/component';
+ import { Suspense } from 'react';
+ import { loadable } from 'sku/vite/loadable';

- const Home = loadable(() => import('./Home'), {
-   fallback: <div>Loading Home...</div>,
- });
+ const Home = loadable(() => import('./Home'));

export default () => (
  <div>
+   <Suspense fallback={<div>Loading Home...</div>}>
    <Home />
+   </Suspense>
  </div>
);
```

### Using vite types

If `vite/client` reference types are needed in your project, you can import them from `sku/vite/client`.

```typescript
/// <reference types="sku/vite/client" />

// ... rest of your application

> ````

Reviewer: ````suggestion
```
````
